haskell-modules/generic-builder.nix: Allow running Haskell tests as passthru.tests

TeofilC opened the pull request:
## Description of changes

This adds a flag to the Haskell generic builder to allow running the test suites through a `passthru.tests.checks` derivation.

This is quite helpful for large/complex test suites attached to large libraries. Having these as separate derivations makes it a lot quicker to debug why a test suite is failing. It's also quite nice for end-users cause there's a clearer separation between "this build failed" and "this test-suite failed".

It also means that if *just* the test suite fails then the overall build graph isn't blocked. So you can potentially build more packages and get more information from one feedback cycle. This makes a big difference when building large package graphs.

If a test requires KVM, this is also helpful for just making the test derivation require that.

This is a feature in haskell.nix and I'm taking inspiration from their implementation. (I'll add credit in the commit message once I write up a proper one). For reference here's the equivalent bit of haskell.nix: https://github.com/input-output-hk/haskell.nix/blob/master/lib/check.nix, which has a few more bells and whistles.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

maralorn commented:
I like the idea of this, but I think we need a bit more discussion on the design.

I am not certain whether this design as is will make anything easier. Concretely can you name a kind of iteration on the tests which will not invalidate building the original package?

TeofilC commented:
I'll try to pick this back up and integrate folks comments soon.

Given that we don't want to depend on `testTargets`, I'll make it so that if `testTargets` is empty then we will ask `./Setup` for the targets. This means that we will have one deriviation for all the tests but that seems OK

TeofilC committed:
haskell-modules/generic-builder.nix: Allow running Haskell tests as passthru.tests

We add an `enableSeperateTestOutput` option to the Haskell builder.
When enabled, we install the test suites into a `tests` output and
create a passthru tests.checks deriviation that will run the test suites.

We take the list of test suites from `testTargets` if it is defined, and
probe Cabal for all the test suites otherwise.

TeofilC commented:
Thanks for your comments @maralorn and @wolfgangwalther. I have done some more work on this. I have reworked this to not depend on the list of test suites at eval time.

Instead of having one passthru deriviation for each test suite, we have just one that runs all the test suites. And if no `testTargets` are specified then we simply probe `Setup.hs` to get the list of test suite exes.

I hope this should address your comments. Let me know if you have some more thoughts.

I haven't tested this in depth yet or added new utilities to `haskell.lib...`. I'm hoping to get some feedback about whether this PR is likely to be accepted with this new direction before putting more time into it.

TeofilC commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -565,7 +570,9 @@ stdenv.mkDerivation ({
-  checkPhase = ''
+  # Unless `enableSeparateTestOutput` is set, in which case the test suites are
+  # run in a separate derivation in `passthru.test`.
+  ${if enableSeparateTestOutput then null else "checkPhase"} = ''
```
`doCheck` is about whether we want to enable test suites at configure time. I have simply removed this line instead. While strange, a user could conceivably want to both run the checks in the main derivation and the separate one. So if they don't want it run in the main checkPhase then they should disable that separately. I'm not 100% sure if this is the right way to go, but it seems like the option that makes the fewest assumptions

TeofilC commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -590,6 +597,7 @@ stdenv.mkDerivation ({
   installPhase = ''
     runHook preInstall
 
+    mkdir $out
```
I've deleted this. I can't recall why I needed it. Maybe I'll figure it out when testing this PR

TeofilC commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -691,6 +706,22 @@ stdenv.mkDerivation ({
+            runHook preCheck
+            ${drv.test}/bin/${testExe} $checkFlags ${lib.concatStringsSep " " testFlags} 2> >(tee $out/test-stderr) | tee $out/test-stdout
+            runHook postCheck
+          '';
```
I've applied this suggestion

TeofilC commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -691,6 +706,22 @@ stdenv.mkDerivation ({
+      (lib.genAttrs (lib.splitString " " testTarget) (testExe:
+        stdenv.mkDerivation {
+          name = "${pname}-${testExe}-check";
+          inherit src preCheck postCheck;
```
Good call. I've done that

TeofilC commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -633,6 +641,13 @@ stdenv.mkDerivation ({
 
+    ${optionalString enableSeparateTestOutput ''
+    mkdir -p $test/bin
+    for t in ${testTarget}; do
```
I've updated this

maralorn reviewed (commented):
I think I am convinced this could be useful and am in favor of moving forward.

I still feel that some tests will fail in this rudimentary separate derivation because of a multitude of unexpected reasons. But we will have to work on that iteratively when we encounter it.

Besides that I still feel like we should try our bests to decouple the derivations as much as possible. e.g. with doCheck = false we possibly need to null the checkPhases stuff in the original derivation so that modifying it does not invalidate the derivation?

Would it be possible to do the building of the tests also in the test derivation?

maralorn commented:
Appears sensible to me. Have you already experimented on a few packages whether the separate test works?

TeofilC commented:
I rebased and tested this a bit. It looks like the behaviour when testTargets isn't set is quite flakey. It only works with certain versions of `Cabal`. I can either make it more robust or just only allow this if the user is also using testTargets
